src/query/service/src/catalogs/catalog_manager.rs

```diff
@@ -80,4 +87,47 @@ impl CatalogManagerHelper for CatalogManager {
         }
         Ok(())
     }
+
+    fn create_user_defined_catalog(&self, req: CreateCatalogReq) -> Result<()> {
+        let catalog_type = req.meta.catalog_type;
+
+        // create catalog first
+        match catalog_type {
+            CatalogType::Default => Err(ErrorCode::CreateUnsupportedCatalog(
+                "Creating a DEFAULT catalog is not allowed",
+            )),
+            CatalogType::Hive => {
+                #[cfg(not(feature = "hive"))]
+                {
+                    Err(ErrorCode::CreateUnsupportedCatalog(
+                        "Hive catalog support is not enabled in your databend-query distribution.",
+                    ))
+                }
+                #[cfg(feature = "hive")]
+                {
+                    let catalog_options = req.meta.options;
+                    let hms_address = catalog_options
+                        .get("hms_address")
+                        .ok_or_else(|| ErrorCode::InvalidArgument("Expected HMS_ADDRESS"))?;
+                    let catalog: Arc<dyn Catalog> = Arc::new(HiveCatalog::try_create(hms_address)?);
+                    let ctl_name = req.name_ident.catalog_name.clone();
+                    match self.catalogs.entry(ctl_name.clone()) {
+                        dashmap::mapref::entry::Entry::Vacant(v) => {
```

Comment:
Too deep in dashmap. Why not `get_mut` instead?

Reply:
Concurrent writings will occur here, so this needs to be carefully dealt with.

DashMap organizes data with different `shard`s, and all reading and writings happen on those `shards`. Like a vector of `RwLock`s.

Both `DashMap::get_mut` and `DashMap::entry` will acquire a write lock on the shard, but if the entry happened to be vacant:

- `DashMap::get_mut` will return `None` and then the shard is `unlocked`, see [this](https://docs.rs/dashmap/latest/src/dashmap/lib.rs.html#986-1006) for detail;
- `DashMap::entry` will return a `VacantEntry`, which occupies the write lock on the shard, see [this](https://docs.rs/dashmap/latest/src/dashmap/lib.rs.html#1105-1121) and [this](https://docs.rs/dashmap/latest/src/dashmap/mapref/entry.rs.html#87-90) for detail;

Surely you can see, using `DashMap::get_mut` while ensuring concurrent safety is more complex than `DashMap::entry`.

Reply:
Lesson learned. Please wrap this function in `CatalogManager` with the detailed explanation so that we don't need to learn this every time.